### PR TITLE
Initial thoughts for parent propagation (WIP)

### LINKED ISSUE #156: Default span propagation for tracing in builders
Following on from https://github.com/jtescher/opentelemetry-rust/issues/17 

It would be useful if the server and client builders provided a default implementing for propagating parent spans via the `grpc-trace-bin` header. Currently I have something similar to this for a service I'm working on using https://github.com/census-instrumentation/opencensus-go/blob/17d7955af9d42886455ce010dd46878208041a58/plugin/ocgrpc/trace_common.go as a reference for how it's typically done in other languages:

```Rust
Server::builder()
    .interceptor_fn(|svc, req| {
        let grpc_trace = &req.headers()["grpc-trace-bin"];
        let bp = BinaryPropagator::new();
        let span_state = bp.from_bytes(grpc_trace.as_bytes().into().span_id();
        let parent = match span_state {
            0 => None,
            e @ _ => Some(Id::from_u64(e)),
        };
        let root = match req.uri().path() {
            "/package.service/method" => span!(parent: parent, Level::INFO, "package.service.method"),
            _ => span!(parent: parent, Level::INFO, "package.service.unknown"),
        };
        let _enter = root.enter();
        svc.call(req)
    })
    .add_service(service).serve(addr).await?;
```

The ugly match is because the span name needs a static lifetime so if there was also a static string representing the `package.service.name` identifier in the generated grpc code that would be useful as well.

### PR DESCRIPTION
**Pull request by xd009642**
## Motivation

Very initial sketch of what a grpc parent propagation may look like. Will hopefully fix #156 eventually

## Solution

Use the interception function to create a parent span. Unfortunately the span binary can differ from tracing system to tracing system so I've kept this initially very very general. No solution for good span names, very WIP.

### CODE REVIEW

**Comment by xd009642** [comment]
Also it would be super helpful if there was a way to put non-static strings into span names though I do have a vague inkling on why that's not possible

File: tonic/src/transport/server.rs [hunk 1]
```diff
@@ -173,6 +175,24 @@ impl Server {
...
+    pub fn propagate_trace_parents<F>(self, f: F) -> Self
+    where
+        F: Fn(&HeaderValue) -> Option<Id> + Send + Sync + 'static,
+    {
+        self.interceptor_fn(move |svc, req| {
+            if let Some(grpc_trace) = req.headers().get("grpc-trace-bin") {
+                let span_id = f(grpc_trace);
+                let root = span!(parent: span_id, Level::INFO, "request");
+                let _enter = root.enter();
+                svc.call(req)
```
Comment (davidbarsky) [inline tonic/src/transport/server.rs]: This won't be propagated beyond this span, unfortunately. `_enter` would dropped at the end of this scope. You'd need something like: https://github.com/tokio-rs/tracing/blob/master/tracing-tower/src/service_span.rs, which hasn't been updated to the releases of Tower using `std::future::Future`. That might happen this week or next week...?

Comment (xd009642) [reply tonic/src/transport/server.rs]: Probably a noob question but is it possible to use `tracing-futures` to do this by `svc.call(req).instrument(root)` ?

Comment (davidbarsky) [reply tonic/src/transport/server.rs]: I'm not 100% sure, but I think the instrumentation needs to be on the Service-level. Give it a shot?

Comment (hawkw) [reply tonic/src/transport/server.rs]: > Probably a noob question but is it possible to use `tracing-futures` to do this by `svc.call(req).instrument(root)` ?

Yeah, that will work fine!

(@davidbarsky, this is all that `tracing-tower` does :P)

**Comment by hawkw** [comment]
> Also it would be super helpful if there was a way to put non-static strings into span names though I do have a vague inkling on why that's not possible

Span _names_ have to be string literals, but you can put an arbitrary non-static string in a span _field_, is that sufficient?

For example:
```rust
let mut my_string = String::new();
my_string.push_str("this string");
my_string.push_str(" is constructed");
my_string.push_str(" dynamically");
span!(Level::INFO, "my_span_name", my_string)
```

File: tonic/src/transport/server.rs (hunk 1 above)
Comment (hawkw) [reply tonic/src/transport/server.rs]: I think allowing users to allow their own span is probably a good idea.

**Comment by davidbarsky** [comment]
@hawkw 

> Yeah, that will work fine!
> (@davidbarsky, this is all that tracing-tower does :P)

oops lol

**Pull request closed without merging**